File: tests/r_parity/05_sunab.py
"""StatsPAI Sun-Abraham event-study parity (Python side) -- Module 05.

Runs sp.sun_abraham(..., aggregation="fixest_att") on the mpdta
replica and emits the R/Stata-compatible weighted average ATT plus
the dynamic event-study coefficients at relative times -3, -2, 0, 1,
2 (post-period anchor reference). The
companion R script runs fixest::feols(... | sunab(g, t)) on the
same CSV.

Tolerance: rel < 1e-3 for the simple weighted-average ATT; the
event-study coefficients are reported as a sanity-check.
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "05_sunab"


def main() -> None:
    df = sp.datasets.mpdta()
    dump_csv(df, MODULE)

    fit = sp.sun_abraham(
        df, y="lemp", g="first_treat", t="year", i="countyreal",
        aggregation="fixest_att",
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE, side="py", statistic="weighted_avg_ATT",
            estimate=float(fit.estimate),
            se=float(fit.se),
            ci_lo=float(fit.ci[0]) if fit.ci is not None else None,
            ci_hi=float(fit.ci[1]) if fit.ci is not None else None,
            n=int(len(df)),
        )
    ]
    rows.append(
        ParityRecord(
            module=MODULE, side="py", statistic="event_time_avg_ATT",
            estimate=float(fit.model_info["att_event_time"]),
            se=float(fit.model_info["se_event_time"]),
            n=int(len(df)),
        )
    )

    # Per-relative-time event-study coefficients.
    es = fit.model_info.get("event_study")
    if es is not None:
        for _, row in es.iterrows():
            rt = int(row["relative_time"])
            rows.append(
                ParityRecord(
                    module=MODULE, side="py",
                    statistic=f"att_rel_{rt}",
                    estimate=float(row["att"]),
                    se=float(row["se"]),
                    ci_lo=float(row["ci_lower"]),
                    ci_hi=float(row["ci_upper"]),
                    n=int(len(df)),
                )
            )

    write_results(
        MODULE, "py", rows,
        extra={
            "control_group": "nevertreated",
            "method": fit.method,
            "aggregation": fit.model_info["summary_aggregation"],
            "aggregation_parity_note": (
                "The weighted_avg_ATT row uses "
                "sp.sun_abraham(..., aggregation='fixest_att'), which "
                "weights post-treatment cohort-time cells by treated "
                "cohort size and matches fixest::summary(..., agg='att') "
                "on the mpdta fixture. The historical equal-weighted "
                "post-event-time summary is retained as event_time_avg_ATT."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/05_sunab.R
# StatsPAI Sun-Abraham parity (R side) -- Module 05.
#
# Reads data/05_sunab.csv (the StatsPAI mpdta replica) and runs
# fixest::feols(lemp ~ sunab(first_treat, year) | countyreal + year).
# Tolerance: rel < 1e-3 (iterative).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(fixest)
})

MODULE <- "05_sunab"

df <- read_csv_strict(MODULE)
df$first_treat <- as.numeric(df$first_treat)

fit <- fixest::feols(
  lemp ~ sunab(first_treat, year) | countyreal + year,
  data = df,
  cluster = ~ countyreal
)

# Aggregated weighted-average ATT (post-treatment, all cohorts).
agg <- summary(fit, agg = "att")
agg_co <- coef(agg)
agg_se <- sqrt(diag(vcov(agg)))

# Locate the aggregated post-treatment row.
post_name <- grep("^ATT$|post|Post|^after$", names(agg_co), value = TRUE)
if (length(post_name) == 0) {
  # fixest names the aggregated row variously across versions; fall
  # back to the first coefficient.
  post_name <- names(agg_co)[1]
}

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "weighted_avg_ATT",
    estimate  = unname(agg_co[post_name]),
    se        = unname(agg_se[post_name]),
    n         = nrow(df)
  )
)

# Event-study coefficients: extract from the disaggregated summary.
es <- summary(fit)
co <- coef(es)
se <- sqrt(diag(vcov(es)))
# fixest's sunab() names the coefficients like "year::-4:cohort::2004"
# OR returns a single column per relative time (e.g. when use the
# "att" agg the names go away). To match per-rel-time, we re-run
# without the agg.
es2 <- summary(fit, agg = FALSE)
co2 <- coef(es2)
se2 <- sqrt(diag(vcov(es2)))
# Each name has form "year::<rel>:cohort::<g>" -- aggregate over
# cohorts using sunab's weighted aggregation by hand to mirror
# StatsPAI's per-rel-time output.
parse_rel <- function(nm) {
  m <- regmatches(nm, regexec("year::(-?\\d+):cohort::(\\d+)", nm))
  if (length(m[[1]]) != 3L) return(NA_integer_)
  as.integer(m[[1]][2])
}
rels <- sapply(names(co2), parse_rel)
keep <- !is.na(rels) & rels != -1L
co2 <- co2[keep]; se2 <- se2[keep]; rels <- rels[keep]

# Cohort-share weighted average per relative time.
# Get cohort counts from the data.
treated <- subset(df, first_treat > 0)
cohort_n <- table(treated$first_treat) / max(treated$first_treat - treated$first_treat + 1)
cohort_w <- cohort_n / sum(cohort_n)

agg_per_rel <- function(rt) {
  m <- regmatches(names(co2), regexec("year::(-?\\d+):cohort::(\\d+)", names(co2)))
  ms <- do.call(rbind, lapply(m, function(x) if (length(x) == 3L) c(x[2], x[3]) else c(NA, NA)))
  rel_match <- as.integer(ms[, 1]) == rt & !is.na(ms[, 1])
  if (!any(rel_match)) return(c(NA, NA))
  cohorts <- as.character(as.integer(ms[rel_match, 2]))
  est <- co2[rel_match]
  s   <- se2[rel_match]
  w   <- cohort_w[cohorts]
  if (any(is.na(w))) {
    return(c(weighted_average_unequal_weights = NA, NA))
  }
  w <- w / sum(w)
  c(sum(w * est), sqrt(sum(w^2 * s^2)))
}

for (rt in sort(unique(rels))) {
  out <- agg_per_rel(rt)
  rows[[length(rows) + 1L]] <- parity_row(
    module    = MODULE,
    statistic = paste0("att_rel_", rt),
    estimate  = out[1],
    se        = out[2],
    n         = nrow(df)
  )
}

write_results(MODULE, rows, extra = list(cluster = "countyreal"))
